cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/MessagingAuditLogWriter.java

```diff
@@ -0,0 +1,122 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.auth;
+
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.auditlogging.AuditLogWriter;
+import io.cdap.cdap.api.messaging.TopicAlreadyExistsException;
+import io.cdap.cdap.api.messaging.TopicNotFoundException;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.messaging.DefaultTopicMetadata;
+import io.cdap.cdap.messaging.client.StoreRequestBuilder;
+import io.cdap.cdap.messaging.spi.MessagingService;
+import io.cdap.cdap.messaging.spi.StoreRequest;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.TopicId;
+import io.cdap.cdap.security.spi.authorization.AuditLogContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Queue;
+import java.util.Random;
+
+/**
+ * This class receives a collection of {@link AuditLogContext} and writes them in order to a
+ * messaging service / topic  ( ex - tms )
+ */
+public class MessagingAuditLogWriter implements AuditLogWriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MessagingAuditLogWriter.class);
+  private static final Gson GSON = new Gson();
+
+  private final TopicId topic;
+  private final MessagingService messagingService;
+  private final RetryStrategy retryStrategy;
+
+  @Inject
+  public MessagingAuditLogWriter(CConfiguration cConf, MessagingService messagingService) {
+    this(
+      messagingService,
+      cConf.get(Constants.AuditLogging.AUDIT_LOG_EVENT_TOPIC),
+      cConf.getInt(Constants.AuditLogging.AUDIT_LOG_EVENT_NUM_PARTITIONS),
+      RetryStrategies.fromConfiguration(cConf, Constants.AuditLogging.AUDIT_LOG_WRITER_RETRY_PREFIX)
+    );
+  }
+
+  /**
+   * Create a publisher that writes {@link AuditLogContext}s to MessagingService topics.
+   * Every operation / event for a Call should be published to a particular topic.
+   * So, when this class is initialized we randomly pick 1 topic based on AUDIT_LOG_EVENT_NUM_PARTITIONS.
+   */
+  public MessagingAuditLogWriter(MessagingService messagingService,
+                                 String topicPrefix, int numTopics, RetryStrategy retryStrategy) {
+    this.messagingService = messagingService;
+    String topicName = topicPrefix + (new Random().nextInt(numTopics) + 1);
```

Comment:
This would select a single topic per pod. It may create hot topics for more loaded pods (e.g. app fabric). I would select a topic per pod per thread.

E.g. use thread name hash and per-writer fixed random value to determine the topic ID. This would ensure single thread always writes to same topic, thus handling any retries, plus same thread on different pods would write to different topics (with the random value).